torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2276,50 +2278,70 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
-  // elements of the matrix (i.e., the lower triangular portion of the input) and then
-  // reflect it on the upper triangular portion, thereby symmetrizing the gradient of
-  // the symeig operation. The motivation behind this choice is that symmetric gradient
-  // leads to stable gradient updates, and retains symmetry of the updated matrix if it
-  // were updated by a gradient based algorithm.
+// https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
+Tensor eigh_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
+                     bool eigenvectors, const Tensor& lambda, const Tensor& v) {
+  // This function is used for both torch.symeig and torch.linalg.eigh.
+  // eigh (and torch.symeig) operates only on symmetric (resp. Hermitian) inputs.
+  // This is a _linear_ subspace of R^{n x n} (resp. C^{n x n}),
+  // since A, B \in Sym(n) => A + B \in Sym(n)
+  // This means that, the tangent space to Sym(n) is Sym(n),
+  // so the gradient is symmetric (resp. Hermitian).
+  // For this reason, we disregard the `upper` flag in torch.symeig and the `UPLO` flag
+  // in torch.linalg.eigh, and we simply work with symmetric matrices.
+
+  // This check just can be triggered in the backwards of torch.symeig
   TORCH_CHECK(eigenvectors,
-           "symeig_backward: Setting eigenvectors to false in torch.symeig doesn't compute eigenvectors ",
-           "and hence we cannot compute backward. Please use torch.symeig(eigenvectors=True)");
+           "Setting eigenvectors=False in torch.symeig doesn't compute the eigenvectors, ",
+           "and hence we cannot compute the backward. Please use torch.symeig(eigenvectors=True)");
 
   auto glambda = grads[0];
   auto gv = grads[1];
 
   auto vh = v.conj().transpose(-2, -1);
 
+  if (glambda.defined()) {
+    // glambda is real even when the matrix is complex
+    glambda = glambda.to(self.dtype());
+  }
+
   Tensor result;
   if (gv.defined()) {
-      Tensor F = lambda.unsqueeze(-2) - lambda.unsqueeze(-1);
-      F.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(INFINITY);
-      F.pow_(-1);
-      result = at::matmul(v, at::matmul(F * at::matmul(vh, gv), vh));
-  } else {
-      result = at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
-  }
+    // Computes v @ (diag(glambda) + F*(vh @ gv)) @ vh as per the source above
+    Tensor F = lambda.unsqueeze(-2) - lambda.unsqueeze(-1);
+    if (at::GradMode::is_enabled()) {
+      // Avoids differentiating through at infinity when doing gradgrad
+      // 1 could be any number, as we are going to overwrite the diagonal
+      F.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(1);
+    }
+    // We compute first the multiplication F*(vh @ gv) and then we zero-out the diagonal.
+    // Otherwise, if vh @ gh has NaNs on the diagonal, these would be propagated to the result
+    // as NaN * 0 = NaN.
+    result =  at::matmul(vh, gv) / F;
+    if (glambda.defined()) {
```

Comment:
I disagree that the gradient is symmetric, primarily because the matrix `F` is NOT symmetric.

Reply:
F is skew, and `vh @ gv` is skew (or skew-Hermitian), so the result is symmetric. Looking at the derivation you can also prove that its diagonal is zero, which is why we can do the copy below.

Reply:
Hmm, why is that `vh @ gv` is skew-symmetric?

Reply:
If that is the case, what we should do is project `vh @ gv` to be skew-symmetric. In theory, `vh @ gv` should be skew-symmetric since if you differentiate `V^T V = Id` for `V \in R^{n x k}` with orthonormal columns and `n >= k`, you get that `V^T  gv + gv^T V = 0`, that is, `V^T gv` is skew-symmetric.

Reply:
As discussed, yes, `V^T dV` is indeed skew-symmetric, but `dV` and `gv` are completely different beasts. `dV` comes from the forward pass (differential), and `gv` comes from the backward pass (it is a gradient).